okhttp-testing-support/src/main/java/okhttp3/RecordingEventListener.kt

```diff
@@ -63,21 +69,50 @@ open class RecordingEventListener : EventListener() {
    * and returns it.
    */
   fun <T> removeUpToEvent(eventClass: Class<T>): T {
-    val fullEventSequence = eventSequence.toMutableList()
-    var event = eventSequence.poll()
-    while (event != null && !eventClass.isInstance(event)) {
-      event = eventSequence.poll()
+    val fullEventSequence = eventSequence.toList()
+    try {
+      while (true) {
+        val event = takeEvent()
+        if (eventClass.isInstance(event)) {
+          return eventClass.cast(event)
+        }
+      }
+    } catch (e: NoSuchElementException) {
+      throw AssertionError("full event sequence: $fullEventSequence", e)
+    }
+  }
+
+  /**
+   * Remove and return the next event from the recorded sequence.
+   *
+   * @param eventClass a class to assert that the returned event is an instance of, or null to
+   *     take any event class.
+   * @param elapsedMs the time in milliseconds elapsed since the immediately-preceding event, or
+   *     -1L to take any duration.
+   */
+  fun takeEvent(eventClass: Class<*>? = null, elapsedMs: Long = -1L): CallEvent {
+    val result = eventSequence.remove()
+    val actualElapsedNs = result.timestampNs - (lastTimestampNs ?: result.timestampNs)
+    lastTimestampNs = result.timestampNs
+
+    if (eventClass != null) {
+      assertThat(result).isInstanceOf(eventClass)
     }
-    if (event == null) {
-      throw AssertionError("${eventClass.simpleName} not found. Found $fullEventSequence.")
+
+    if (elapsedMs != -1L) {
+      assertThat(TimeUnit.NANOSECONDS.toMillis(actualElapsedNs).toDouble())
+          .isCloseTo(elapsedMs.toDouble(), Offset.offset(100.0))
     }
-    return eventClass.cast(event)
+
+    return result
   }
 
   fun recordedEventTypes() = eventSequence.map { it.name }
 
   fun clearAllEvents() {
-    eventSequence.clear()
+    while (eventSequence.isNotEmpty()) {
+      takeEvent()
+    }
```

Comment:
What's the difference here?

Reply:
I need to call takeEvent() so the `lastTimestampNs` is updated.